### PR TITLE
Attractivity phase

### PR DESCRIPTION
**Pull request by GloeStoe**
implemented the attractivity phase and tests for it

### CODE REVIEW

File: src/osm_configurator/model/project/calculation/attractivity_phase.py [hunk 1]
```diff
@@ -25,4 +55,108 @@ def calculate(self, configuration_manager):
...
+        # Delete result files
+        result_folder: Path = calculation_phase_utility.get_checkpoints_folder_path_from_phase\
+            (configuration_manager, calculation_phase_enum.CalculationPhase.ATTRACTIVITY_PHASE)
+        deleter: FileDeletion = file_deletion.FileDeletion()
+        deleter.reset_folder(result_folder)
+
+        # Iterate over all traffic cells and generate the attractivities
+        index: int
+        row: Series
+        for index, row in cells.iterrows():
```
Comment (LuposX) [inline src/osm_configurator/model/project/calculation/attractivity_phase.py:79]: IDK, i did it differently(iterating over all fiels in a folder) i don't know how hard is it to support doing it both ways.
Maybe need to change it, but @GloeStoe knows if he needs to change it, dependent on how he implements process manager.

File: src/osm_configurator/model/project/calculation/attractivity_phase.py [hunk 2]
```diff
@@ -25,4 +55,108 @@ def calculate(self, configuration_manager):
...
+        output_df.to_csv(output_path)
+
+        return calculation_state_enum.CalculationState.RUNNING, "running"
+
+    def _calculate_attractivity_for_element(self, element: Series, output_df: DataFrame,
+                                            category_list: List[Category]) -> DataFrame:
+        category_name: str = element["category"]
+        category: Category = self._get_category_by_name(category_name, category_list)
+
+        new_entry = pds.Series()
```
Comment (LuposX) [inline src/osm_configurator/model/project/calculation/attractivity_phase.py:126]: typehint

File: src/osm_configurator/model/project/calculation/attractivity_phase.py [hunk 3]
```diff
@@ -25,4 +55,108 @@ def calculate(self, configuration_manager):
...
+            attribute: Attribute
+            for attribute in attribute_enum.Attribute:
+                value += attractivity.get_attribute_factor(attribute) * element[attribute.get_name()]
+            new_entry[attractivity.get_attractivity_attribute_name()] = value
+
+        output_df = output_df.append(new_entry, ignore_index=True)
+        return output_df
+
+    def _get_category_by_name(self, category_name: str, category_list: List[Category]) -> Category:
+        list_of_categories_with_name = [cat for cat in category_list if cat.get_category_name() == category_name]
```
Comment (LuposX) [inline src/osm_configurator/model/project/calculation/attractivity_phase.py:140]: typehint

File: src/osm_configurator/model/project/calculation/attractivity_phase.py [hunk 4]
```diff
@@ -25,4 +55,108 @@ def calculate(self, configuration_manager):
...
+        # Read and create data frames
+        input_df: DataFrame = pd.read_csv(input_path)
+        output_df: DataFrame = self._init_dataframe(config_manager.get_category_manager())
+
+        # Iterate over all elements in this cell
+        index: int
+        element: Series
+        for index, element in input_df.iterrows():
+            try:
+                output_df = self._calculate_attractivity_for_element(element, output_df,
```
Comment (LuposX) [inline src/osm_configurator/model/project/calculation/attractivity_phase.py:111]: This is bad style you shouldn't do this(this leads to evry poor performance), look below at method implementation for details.

File: src/osm_configurator/model/project/calculation/attractivity_phase.py [hunk 5]
```diff
@@ -25,4 +55,108 @@ def calculate(self, configuration_manager):
...
+
+        attractivity: AttractivityAttribute
+        for attractivity in category.get_attractivity_attributes():
+            value: float = attractivity.get_base_factor()
+            attribute: Attribute
+            for attribute in attribute_enum.Attribute:
+                value += attractivity.get_attribute_factor(attribute) * element[attribute.get_name()]
+            new_entry[attractivity.get_attractivity_attribute_name()] = value
+
+        output_df = output_df.append(new_entry, ignore_index=True)
```
Comment (LuposX) [inline src/osm_configurator/model/project/calculation/attractivity_phase.py:136]: Dont use append, its very slow see stackoverflow link below(third answer for how slow it is, second answer on how to do it better).

Instead append to a dictinary and then create fro mthe dictionary a dataframe.

Comparsion between append and from dict:
https://stackoverflow.com/questions/10715965/create-a-pandas-dataframe-by-appending-one-row-at-a-time

Screenshot from 2023-02-07 19-56-08

**Review by LuposX — CHANGES_REQUESTED** [review]
Fix this with how you create dataframe else looks good

File: src/osm_configurator/model/project/calculation/attractivity_phase.py (hunk 1 above)
Comment (GloeStoe) [reply src/osm_configurator/model/project/calculation/attractivity_phase.py:79]: taken the note

File: src/osm_configurator/model/project/calculation/attractivity_phase.py (hunk 2 above)
Comment (GloeStoe) [reply src/osm_configurator/model/project/calculation/attractivity_phase.py:126]: fixed

File: src/osm_configurator/model/project/calculation/attractivity_phase.py (hunk 3 above)
Comment (GloeStoe) [reply src/osm_configurator/model/project/calculation/attractivity_phase.py:140]: fixed

**Review by LuposX — APPROVED** [review]

**Pull request merged**